diesel/src/query_builder/update_statement/mod.rs

```diff
@@ -59,35 +59,53 @@ impl<T, U, DB> QueryFragment<DB> for UpdateStatement<T, U> where
 }
 
 impl<T, U> AsQuery for UpdateStatement<T, U> where
-    UpdateQuery<T, U>: Query,
+    T: UpdateTarget,
+    UpdateQuery<<T::Table as Table>::AllColumns, UpdateStatement<T, U>>: Query,
 {
     type SqlType = <Self::Query as Query>::SqlType;
-    type Query = UpdateQuery<T, U>;
+    type Query = UpdateQuery<<T::Table as Table>::AllColumns, UpdateStatement<T, U>>;
 
     fn as_query(self) -> Self::Query {
-        UpdateQuery(self)
+        UpdateQuery {
+            returning: T::Table::all_columns(),
+            statement: self,
+        }
+    }
+}
+
+impl<T, U> UpdateStatement<T, U> {
+    /// Specify what expression is returned after execution of the `update`.
+    pub fn returning<E>(self, returns: E) -> UpdateQuery<E, UpdateStatement<T, U>> where
+        E: Expression,
+    {
+        UpdateQuery {
+            returning: returns,
+            statement: self,
+        }
     }
 }
 
 #[doc(hidden)]
-pub struct UpdateQuery<T, U>(UpdateStatement<T, U>);
+pub struct UpdateQuery<T, U> {
+    returning: T,
+    statement: U,
+}
 
 impl<T, U, DB> QueryFragment<DB> for UpdateQuery<T, U> where
     DB: Backend + SupportsReturningClause,
-    T: UpdateTarget,
-    <T::Table as Table>::AllColumns: QueryFragment<DB>,
-    UpdateStatement<T, U>: QueryFragment<DB>,
+    T: QueryFragment<DB>,
+    U: QueryFragment<DB>,
 {
     fn to_sql(&self, out: &mut DB::QueryBuilder) -> BuildQueryResult {
-        try!(self.0.to_sql(out));
+        try!(self.statement.to_sql(out));
         out.push_sql(" RETURNING ");
-        try!(T::Table::all_columns().to_sql(out));
+        try!(self.returning.to_sql(out));
         Ok(())
     }
 }
 
 impl<T, U> Query for UpdateQuery<T, U> where
-    T: UpdateTarget,
+    T: Expression,
```

Comment:
This removes a lot of type safety constraints. This needs to continue to verify that we're referencing an update target.